File: packages/cortex-python/cortex_sdk/resources/teams.py
"""Teams resource."""

from __future__ import annotations

from typing import Optional

from cortex_sdk._http import AsyncHTTPClient, SyncHTTPClient
from cortex_sdk.types import (
    DeleteResponse,
    Team,
    TeamCreate,
    TeamList,
    TeamMember,
    TeamMemberAdd,
    TeamMemberUpdate,
)


class Teams:
    """Synchronous teams resource."""

    def __init__(self, http: SyncHTTPClient) -> None:
        self._http = http

    def list(self) -> TeamList:
        data = self._http.request("GET", "/api/teams")
        return TeamList.model_validate(data)

    def create(self, *, name: str, description: Optional[str] = None) -> Team:
        req = TeamCreate(name=name, description=description)
        data = self._http.request("POST", "/api/teams", json=req.model_dump(exclude_none=True))
        return Team.model_validate(data)

    def get(self, team_id: str) -> Team:
        data = self._http.request("GET", f"/api/teams/{team_id}")
        return Team.model_validate(data)

    def delete(self, team_id: str) -> DeleteResponse:
        data = self._http.request("DELETE", f"/api/teams/{team_id}")
        return DeleteResponse.model_validate(data)

    def add_member(
        self,
        team_id: str,
        *,
        email: str,
        role: Optional[str] = "member",
    ) -> TeamMember:
        req = TeamMemberAdd(email=email, role=role)
        data = self._http.request(
            "POST",
            f"/api/teams/{team_id}/members",
            json=req.model_dump(exclude_none=True),
        )
        return TeamMember.model_validate(data)

    def update_member(self, team_id: str, member_id: str, *, role: str) -> TeamMember:
        req = TeamMemberUpdate(role=role)
        data = self._http.request(
            "PATCH",
            f"/api/teams/{team_id}/members/{member_id}",
            json=req.model_dump(),
        )
        return TeamMember.model_validate(data)

    def remove_member(self, team_id: str, member_id: str) -> DeleteResponse:
        data = self._http.request("DELETE", f"/api/teams/{team_id}/members/{member_id}")
        return DeleteResponse.model_validate(data)


class AsyncTeams:
    """Asynchronous teams resource."""

    def __init__(self, http: AsyncHTTPClient) -> None:
        self._http = http

    async def list(self) -> TeamList:
        data = await self._http.request("GET", "/api/teams")
        return TeamList.model_validate(data)

    async def create(self, *, name: str, description: Optional[str] = None) -> Team:
        req = TeamCreate(name=name, description=description)
        data = await self._http.request(
            "POST", "/api/teams", json=req.model_dump(exclude_none=True)
        )
        return Team.model_validate(data)

    async def get(self, team_id: str) -> Team:
        data = await self._http.request("GET", f"/api/teams/{team_id}")
        return Team.model_validate(data)

    async def delete(self, team_id: str) -> DeleteResponse:
        data = await self._http.request("DELETE", f"/api/teams/{team_id}")
        return DeleteResponse.model_validate(data)

    async def add_member(
        self,
        team_id: str,
        *,
        email: str,
        role: Optional[str] = "member",
    ) -> TeamMember:
        req = TeamMemberAdd(email=email, role=role)
        data = await self._http.request(
            "POST",
            f"/api/teams/{team_id}/members",
            json=req.model_dump(exclude_none=True),
        )
        return TeamMember.model_validate(data)

    async def update_member(self, team_id: str, member_id: str, *, role: str) -> TeamMember:
        req = TeamMemberUpdate(role=role)
        data = await self._http.request(
            "PATCH",
            f"/api/teams/{team_id}/members/{member_id}",
            json=req.model_dump(),
        )
        return TeamMember.model_validate(data)

    async def remove_member(self, team_id: str, member_id: str) -> DeleteResponse:
        data = await self._http.request("DELETE", f"/api/teams/{team_id}/members/{member_id}")
        return DeleteResponse.model_validate(data)

File: packages/cortex-dart/lib/src/resources/teams.dart
/// Teams management resource.
library;

import '../errors.dart';
import '../http_client.dart';
import '../types.dart';

/// Provides access to team management endpoints.
class TeamsResource {
  final CortexHttpClient _client;
  final String _baseUrl;

  /// Creates a [TeamsResource].
  TeamsResource({
    required CortexHttpClient client,
    required String baseUrl,
  })  : _client = client,
        _baseUrl = baseUrl;

  /// Lists all teams.
  Future<List<Team>> list() async {
    final json = await _client.get('$_baseUrl/api/teams');
    final data =
        json['data'] as List<dynamic>? ?? json['teams'] as List<dynamic>?;
    if (data != null) {
      return data
          .map((e) => Team.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return <Team>[];
  }

  /// Creates a new team.
  ///
  /// [name] is the team name.
  /// [description] is an optional description.
  Future<Team> create({
    required String name,
    String? description,
  }) async {
    if (name.isEmpty) {
      throw const CortexValidationException(
        message: 'Name must not be empty.',
        parameter: 'name',
      );
    }

    final body = <String, dynamic>{'name': name};
    if (description != null) body['description'] = description;

    final json = await _client.post('$_baseUrl/api/teams', body: body);
    return Team.fromJson(json);
  }

  /// Gets team details by ID.
  Future<Team> get(String id) async {
    if (id.isEmpty) {
      throw const CortexValidationException(
        message: 'ID must not be empty.',
        parameter: 'id',
      );
    }
    final json = await _client.get('$_baseUrl/api/teams/$id');
    return Team.fromJson(json);
  }

  /// Deletes a team.
  Future<void> delete(String id) async {
    if (id.isEmpty) {
      throw const CortexValidationException(
        message: 'ID must not be empty.',
        parameter: 'id',
      );
    }
    await _client.delete('$_baseUrl/api/teams/$id');
  }

  /// Adds a member to a team.
  ///
  /// [teamId] is the team identifier.
  /// [email] is the member's email address.
  /// [role] is the role to assign (e.g., "member", "admin").
  Future<TeamMember> addMember({
    required String teamId,
    required String email,
    required String role,
  }) async {
    if (teamId.isEmpty) {
      throw const CortexValidationException(
        message: 'Team ID must not be empty.',
        parameter: 'teamId',
      );
    }
    if (email.isEmpty) {
      throw const CortexValidationException(
        message: 'Email must not be empty.',
        parameter: 'email',
      );
    }
    if (role.isEmpty) {
      throw const CortexValidationException(
        message: 'Role must not be empty.',
        parameter: 'role',
      );
    }

    final json = await _client.post(
      '$_baseUrl/api/teams/$teamId/members',
      body: {'email': email, 'role': role},
    );
    return TeamMember.fromJson(json);
  }

  /// Updates a team member's role.
  ///
  /// [teamId] is the team identifier.
  /// [memberId] is the member identifier.
  /// [role] is the new role.
  Future<TeamMember> updateMemberRole({
    required String teamId,
    required String memberId,
    required String role,
  }) async {
    if (teamId.isEmpty) {
      throw const CortexValidationException(
        message: 'Team ID must not be empty.',
        parameter: 'teamId',
      );
    }
    if (memberId.isEmpty) {
      throw const CortexValidationException(
        message: 'Member ID must not be empty.',
        parameter: 'memberId',
      );
    }
    if (role.isEmpty) {
      throw const CortexValidationException(
        message: 'Role must not be empty.',
        parameter: 'role',
      );
    }

    final json = await _client.patch(
      '$_baseUrl/api/teams/$teamId/members/$memberId',
      body: {'role': role},
    );
    return TeamMember.fromJson(json);
  }

  /// Removes a member from a team.
  Future<void> removeMember({
    required String teamId,
    required String memberId,
  }) async {
    if (teamId.isEmpty) {
      throw const CortexValidationException(
        message: 'Team ID must not be empty.',
        parameter: 'teamId',
      );
    }
    if (memberId.isEmpty) {
      throw const CortexValidationException(
        message: 'Member ID must not be empty.',
        parameter: 'memberId',
      );
    }
    await _client.delete(
      '$_baseUrl/api/teams/$teamId/members/$memberId',
    );
  }
}
